### PR TITLE
Add discrete_range_rng(lower, upper) and discrete_range_lpmf

### LINKED ISSUE #719: Add discrete uniform rng for stan#2457
#### Summary:
Add a discrete uniform rng for https://github.com/stan-dev/stan/issues/2457.

#### Additional Information:
Can use boost.

#### Current Version:
v2.17.0

### PR DESCRIPTION
**Pull request by mcol**
## Summary

This adds support for new functions `discrete_range_rng(lower, upper)` and `discrete_range_lpmf(y, lower, upper)`, which allow to return a random discrete uniform variate between lower an upper (both extremes inclusive) and its log PMF. Fixes #719.

## Tests

Added. As far as I understand, there's no need to enforce `lower` and `upper` to be integers or integer arrays here, as that will be done when we expose the signatures, so tests don't include the case for which the arguments are not integers.

## Side Effects

None.

## Checklist

- [X] Math issue #719

- [X] Copyright holder: Marco Colombo

    The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [X] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - dependencies checks pass, (`make test-math-dependencies`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Code-Quality) checks (`make cpplint`)

- [X] the code is written in idiomatic C++ and changes are documented in the doxygen

- [X] the new changes are tested

### CODE REVIEW

File: stan/math/prim/prob/discrete_range_rng.hpp [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+ * Any non-scalar inputs must be the same size.
+ *
+ * @tparam T_lower type of lower bound
+ * @tparam T_upper type of upper bound
+ * @tparam RNG type of random number generator
+ *
+ * @param lower lower bound
+ * @param upper upper bound
+ * @param rng random number generator
+ * @return A (sequence of) discrete random variate(s) between `lower` and
```
Comment (bob-carpenter) [inline stan/math/prim/prob/discrete_range_rng.hpp]: The return needs to be a single integer when both arguments are integers, and a standard vector of integers otherwise.  Is that what this is supposed to mean?  Is it how the other RNGs are documnted?  I find it a bit confusing, but if it's like all the other ones, we should go with consistency for now.

File: stan/math/prim/prob/discrete_range_rng.hpp [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+ * `upper`, both bounds included.
+ * @throw std::domain_error if lower or upper are non-finite, or if upper
+ * is smaller than lower
+ * @throw std::invalid_argument if non-scalar arguments are of different
+ * sizes
+ */
+template <typename T_lower, typename T_upper, class RNG>
+inline typename VectorBuilder<true, int, T_lower, T_upper>::type
+discrete_range_rng(const T_lower& lower, const T_upper& upper, RNG& rng) {
```
Comment (bob-carpenter) [inline stan/math/prim/prob/discrete_range_rng.hpp]: The documentation needs to indicate what `T_lower` and `T_upper` can be (presumably either `int` or `std::vector<int>`).  

It should not be possible to compile an instantiation where the lower or upper bounds are `double`.  It's OK if that's enforced by not being able to compile someting inside---it's not like we have to guard against ambiguity.

File: stan/math/prim/prob/discrete_range_rng.hpp [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+  using boost::random::uniform_int_distribution;
+  using boost::variate_generator;
+
+  static const char* function = "discrete_range_rng";
+
+  check_finite(function, "Lower bound parameter", lower);
```
Comment (bob-carpenter) [inline stan/math/prim/prob/discrete_range_rng.hpp]: The `check_finite` tests are not necessary because all integers are finite.

File: stan/math/prim/prob/discrete_range_rng.hpp [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+#include <stan/math/prim/err.hpp>
+#include <stan/math/prim/fun/max_size.hpp>
+#include <boost/random/uniform_int_distribution.hpp>
+#include <boost/random/variate_generator.hpp>
+
+namespace stan {
+namespace math {
+
+/** \ingroup prob_dists
+ * Return a discrete random variate for the given lower and upper bounds
```
Comment (bob-carpenter) [inline stan/math/prim/prob/discrete_range_rng.hpp]: I think it'd help to say that it returns integer(s) between the lower and upper bound(s) inclusive and then elaborate on the return type in the longer doc.  As is, it's very unclear from reading this doc what the return type actually is.   I imagine a lot of the other RNG doc probaby has this same problem.

File: test/unit/math/prim/prob/discrete_range_test.cpp [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+  double nan = std::numeric_limits<double>::quiet_NaN();
+  double inf = std::numeric_limits<double>::infinity();
+
+  Eigen::VectorXd lower(3);
+  Eigen::VectorXd upper(3);
+
+  lower << 5.0, 11.0, -15.0;
+  upper << 5.0, 15.0, -10.0;
+  EXPECT_NO_THROW(discrete_range_rng(lower, upper, rng));
```
Comment (bob-carpenter) [inline test/unit/math/prim/prob/discrete_range_test.cpp]: This should be illegal, so it looks like that Boost thing lets integers get assigned.  So the signature should prevent this with an enable_if on the template parameters.  The tests need to be updated then to use integers.

File: test/unit/math/prim/prob/discrete_range_test.cpp [hunk 6]
```diff
@@ -0,0 +1,73 @@
+#include <stan/math/prim.hpp>
+#include <boost/math/distributions.hpp>
+#include <boost/random/mersenne_twister.hpp>
+#include <gtest/gtest.h>
+#include <limits>
+#include <vector>
+
+TEST(ProbDistributionsDiscreteRange, error_check) {
```
Comment (bob-carpenter) [inline test/unit/math/prim/prob/discrete_range_test.cpp]: The tests need to consider boundary conditions like (lower = 17, upper = 17) and (lower = 17, upper = 18) to make sure we get the right boundary behavior.

File: test/unit/math/prim/prob/discrete_range_test.cpp [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+  boost::random::mt19937 rng;
+
+  int N = 10000;
+  int lower = -3;
+  int upper = 20;
+
+  int K = upper - lower + 1;
+  boost::math::chi_squared mydist(K - 1);
+
+  int bin[K];
```
Comment (bob-carpenter) [inline test/unit/math/prim/prob/discrete_range_test.cpp]: I'd prefer std::vector here with initializers.  E.g., 

```
std::vector<int> bin(K, 0);  //  initialize bins to 0
std::vector<double> expect(K, static_cast<double>(N) / K);
```

File: test/unit/math/prim/prob/discrete_range_test.cpp [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+  double expect[K];
+  double prop = static_cast<double>(N) / K;
+  for (int i = 0; i < K; i++) {
+    bin[i] = 0;
+    expect[i] = prop;
+  }
+
+  int count = 0;
```
Comment (bob-carpenter) [inline test/unit/math/prim/prob/discrete_range_test.cpp]: This should be  a for loop:

```
for (int count = 0; count < N; ++count) {  ... }
```

File: test/unit/math/prim/prob/discrete_range_test.cpp [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+  while (count < N) {
+    int a = stan::math::discrete_range_rng(lower, upper, rng);
+    bin[a - lower]++;
```
Comment (bob-carpenter) [inline test/unit/math/prim/prob/discrete_range_test.cpp]: The `++` should come before the variable if the variable before the increment's not being used.

File: test/unit/math/prim/prob/discrete_range_test.cpp (hunk 5 above)
Comment (mcol) [reply test/unit/math/prim/prob/discrete_range_test.cpp]: No other distribution uses any `enable_if` or `require_` on the template parameters or elsewhere. So my understanding was that this is  enforced by the signatures we expose in the compiler. I can try to concoct something, but it'd be a first in the codebase. I'll update the tests.

File: stan/math/prim/prob/discrete_range_rng.hpp (hunk 2 above)
Comment (mcol) [reply stan/math/prim/prob/discrete_range_rng.hpp]: I've updated the docs. See my other answer as for the instantiation with non-int arguments.

**Comment by mcol** [comment]
Thanks @bob-carpenter for the review. As for restricting instantiations to int or ints[], I have applied no changes as no other probability function does that. If you think it's really necessary, could you show me how to do it? I don't think we have `require_` for int or int array/vectors.

I went ahead and added the lpmf here, following what is done in `hypergeometric_lpmf`, which is the only other function with all int arguments. Since it's quite new territory for me, this deserves extra attention.

File: stan/math/prim/prob/discrete_range_lpmf.hpp [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+#include <stan/math/prim/fun/max_size.hpp>
+#include <stan/math/prim/fun/size_zero.hpp>
+#include <stan/math/prim/fun/value_of.hpp>
+#include <cmath>
+
+namespace stan {
+namespace math {
+
+/** \ingroup prob_dists
+ * Return the log PMF of a discrete range for the given y, lower and upper
```
Comment (bob-carpenter) [inline stan/math/prim/prob/discrete_range_lpmf.hpp]: The documentation should mention that only integers and integer arrays are legal arguments.  This doesn't necessarily have to be enforced anywhere by the function, but it should be documented.

File: stan/math/prim/prob/discrete_range_lpmf.hpp [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+ * bound.
+ *
+ * `lower` and `upper` can each be a scalar or a one-dimensional container.
+ * Any non-scalar inputs must be the same size.
```
Comment (bob-carpenter) [inline stan/math/prim/prob/discrete_range_lpmf.hpp]: [optional]
I prefer "container" to "non-scalar" and "argument" to "input".

File: stan/math/prim/prob/discrete_range_lpmf.hpp [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+#include <stan/math/prim/fun/value_of.hpp>
+#include <cmath>
+
+namespace stan {
+namespace math {
+
+/** \ingroup prob_dists
+ * Return the log PMF of a discrete range for the given y, lower and upper
+ * bound.
+ *
```
Comment (bob-carpenter) [inline stan/math/prim/prob/discrete_range_lpmf.hpp]: This should define the density, which is p(x | lower, upper) = 1 / (upper - lower + 1)` or on the log scale, `log p(x | lower, upper) = -log(upper - lower + 1)`.

File: stan/math/prim/prob/discrete_range_lpmf.hpp [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+ * of the probabilities.
+ * @throw std::domain_error if upper is smaller than lower.
+ * @throw std::invalid_argument if non-scalar arguments are of different
+ * sizes.
+ */
+template <bool propto, typename T_y, typename T_lower, typename T_upper>
+double discrete_range_lpmf(const T_y& y, const T_lower& lower,
+                           const T_upper& upper) {
+  static const char* function = "discrete_range_lpmf";
+
```
Comment (bob-carpenter) [inline stan/math/prim/prob/discrete_range_lpmf.hpp]: [optional]
Vertical whitespace is precious!  I'd get rid of all these extra spaces to allow myself to view more of the function at once.

File: test/unit/math/prim/prob/discrete_range_test.cpp (hunk 5 above)
Comment (bob-carpenter) [reply test/unit/math/prim/prob/discrete_range_test.cpp]: The way to enforce that all arguments are integers or integer arrays, add a final template parameter:

```
typename = std::enable_if_t<std::is_integral<scalar_type_t<T1>>::value
                            && std::is_integral<scalar_type_t<T2>>::value
                            && std::is_integral<scalar_type_t<T3>>::value>
```

where `scalar_type_t` is a Stan meta function.

This isn't strictly necessary, but it should be clear in the doc that the use is intended to only be for integers.

File: stan/math/prim/prob/discrete_range_lpmf.hpp (hunk 13 above)
Comment (mcol) [reply stan/math/prim/prob/discrete_range_lpmf.hpp]: I cut a couple of empty lines. I still prefer keeping logically separate block also separate visually.

**Comment by mcol** [comment]
Thanks Bob! This should be closer to what you expected. :)

File: stan/math/prim/prob/discrete_range_lpmf.hpp (hunk 13 above)
Comment (bob-carpenter) [reply stan/math/prim/prob/discrete_range_lpmf.hpp]: That's why I made it optional.  No promises I won't remove vertical space if I ever touch the files in the future :-)

File: test/unit/math/prim/prob/discrete_range_test.cpp (hunk 5 above)
Comment (bob-carpenter) [reply test/unit/math/prim/prob/discrete_range_test.cpp]: Just to be clear here, our `binomial_lpmf` and our `hypergeometric_lpmf` are broken in that they'll accept real-valued arguments.  I'm going to open a separate issue to try to deal with some of this as it'll be easy to call them erroneously as they're written with real values, but those don't make sense (they don't normalize properly).

File: test/prob/discrete_range/discrete_range_test.hpp [hunk 14]
```diff
@@ -0,0 +1,62 @@
+// Arguments: Ints, Ints, Ints
+#include <stan/math/prim.hpp>
+
+using std::vector;
+
+class AgradDistributionsHypergeometric : public AgradDistributionTest {
```
Comment (bob-carpenter) [inline test/prob/discrete_range/discrete_range_test.hpp]: The name should match what's being tested.  

We should also follow uniform_lpdf and throw if the parameters are not properly ordered.

I believe we should also be throwing a domain error when there's a domain error, that is when the variant isn't within the bounds.  That's not how uniform_lpdf or pareto_lpdf currently work, so I'm going to open another issue to fix those and this one.

I also believe we should be throwing when the variant is not in domain.  That's what we do for other distributions.  The `uniform_lpdf`, but it should be.  

Now that I see these tests, I realize that we should be throwing when there's a domain error, not just returnign -infinity.  For instance, consider `uniform_lpdf`, which throws if the parameters aren't properly ordered.  I'm going to open yet another issue to fix all this in the other distributions.
